watchtower/wtdb/range_index.go

```diff
@@ -0,0 +1,350 @@
+func (a *RangeIndex) AddRange(start, end uint64) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if start > end {
+		return
+	}
+
+	// If the range set is empty, the new range can just be added.
+	if len(a.m) == 0 {
+		a.m[start] = end
+		return
+	}
+
+	// The range is already covered in the index.
+	if num, ok := a.m[start]; ok && end <= num {
+		return
+	}
+
+	// First we assume that the given start value will be the key of the
+	// range we will edit.
+	rangeStart := start
+
+	// Check if there is a range with a start key that is less than the
+	// start key in question.
+	rangeExists, _, rangeBelow := a.highestLowerRange(start)
+	if rangeExists {
+		// If such a range does exist, then we check to see if the
+		// range we are trying to insert is perhaps already covered by
+		// this range.
+		num := a.m[rangeBelow]
+		if end <= num {
+			// The range is already covered in the index.
+			return
+		}
+
+		// Otherwise, check if we can perhaps just extend the existing
+		// range.
+		if num >= start || num+1 == start {
+			// Extend an existing range.
+			rangeStart = rangeBelow
+		}
+	}
+
+	// Write the new range.
+	a.m[rangeStart] = end
+
+	// Update our maxHeight if needed.
+	if end > a.maxHeight {
+		a.maxHeight = end
+	}
+
+	// Now check if we need to merge any another ranges with this one.
+	for {
+		// Find the start value of the lowest range that is still
+		// higher than the start value that we just inserted.
+		rangeExists, rangeAbove := a.lowestHigherRange(rangeStart)
+
+		// If no such range exists, then there is no range to delete.
+		if !rangeExists {
+			return
+		}
+
+		// If there is such a range, but it cannot be merged with the
+		// newly inserted range, then we are done.
+		if rangeAbove > end+1 {
+			return
+		}
+
+		// Otherwise, the two ranges can be merged.
+		num := a.m[rangeAbove]
+		if num >= end {
+			a.m[rangeStart] = num
+		}
+
+		delete(a.m, rangeAbove)
+	}
+}
+
+// IsInIndex returns true if the given number is in the range set.
+func (a *RangeIndex) IsInIndex(n uint64) bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if _, ok := a.m[n]; ok {
+		return true
+	}
+
+	_, isInRange, _ := a.highestLowerRange(n)
+	return isInRange
+}
+
+// NumInSet returns the number of items covered by the range set.
+func (a *RangeIndex) NumInSet() uint64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	var numAcked uint64
+
+	for start, end := range a.m {
+		numAcked += end - start + 1
+	}
+
+	return numAcked
+}
+
+// highestLowerRange finds the largest start value of a range in the set where
+// the start value is still lower than the given value. The first return value
+// is true if such a range exists. The second value returns true if the given
+// number is included in the found range. The last return value is the start
+// value of the found range.
+func (a *RangeIndex) highestLowerRange(h uint64) (bool, bool, uint64) {
+	var (
+		rangeExists   bool
+		rangeStartNum uint64
+	)
+	for height, maxHeight := range a.m {
```

Comment:
This is invoked every time a new height is ACKed. Would it be possible to reverse the iteration order, or implement some other optimization for this use case? I think incrementing the end of the last range is a very common use case, and iterating over all previous ranges to do this doesn't seem fitting. Instead of using a plain map, a "real" data structure (with binary search?) might be more suitable.

Reply:
Maybe this is suitable for a follow-up RP so that the main contribution (with lots of benefits) can be merged faster.

Reply:
am almost done with the improvements 🤓  aiming to push up the changes today

Reply:
done!